### PR TITLE
[Consensus Voting] Implement Single Sig Encoding/Decoding

### PR DESCRIPTION
**Pull request by zhangchiqing**
For https://github.com/dapperlabs/flow-go/issues/5798

### CODE REVIEW

File: consensus/hotstuff/signature/consensus.go [hunk 1]
```diff
@@ -1,14 +1,26 @@
 package signature
 
 import (
+	"fmt"
+
 	"github.com/onflow/flow-go/consensus/hotstuff"
 	"github.com/onflow/flow-go/crypto"
 )
 
 func EncodeSingleSig(sigType hotstuff.SigType, sig crypto.Signature) []byte {
```
Comment (tarakby) [inline consensus/hotstuff/signature/consensus.go:10]:
```suggestion
// EncodeSingleSig encodes a single signature into signature data as required by the consensus design.
func EncodeSingleSig(sigType hotstuff.SigType, sig crypto.Signature) []byte {
```

File: consensus/hotstuff/signature/consensus.go [hunk 2]
```diff
@@ -1,14 +1,26 @@
...
-	panic("to be implemented")
+	t := byte(sigType)
+	encoded := make([]byte, 0, len(sig)+1)
+	encoded = append(encoded, t)
+	encoded = append(encoded, sig[:]...)
+	return encoded
 }
 
 func DecodeSingleSig(sigData []byte) (hotstuff.SigType, crypto.Signature, error) {
```
Comment (tarakby) [inline consensus/hotstuff/signature/consensus.go:18]:
```suggestion
// DecodeSingleSig decodes signature data into a cryptographic signature and a type as required by the consensus design.
func DecodeSingleSig(sigData []byte) (hotstuff.SigType, crypto.Signature, error) {
```

File: consensus/hotstuff/signature/consensus_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func TestEncodeDecodeRandomBeaconSig(t *testing.T) {
+	sig := unittest.SignatureFixture()
+	encoded := EncodeSingleSig(hotstuff.SigTypeRandomBeacon, sig)
+	decodedType, decodedSig, err := DecodeSingleSig(encoded)
+	require.NoError(t, err)
+	require.Equal(t, hotstuff.SigTypeRandomBeacon, decodedType)
+	require.Equal(t, sig, decodedSig)
+}
```
Comment (tarakby) [inline consensus/hotstuff/signature/consensus_test.go:27]: We could merge these 2 tests into one test `TestEncodeDecodeSig` using something like :
```
types := []hotstuff.SigType { hotstuff.SigTypeRandomBeacon, hotstuff.SigTypeStaking} 
for _, t := range types {
...
}
```

File: consensus/hotstuff/signature/consensus_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	require.NotEqual(t, encodedSig1, encodedSig2)
+}
+
+// encode the same sig with the different type, the encoded sig should be different
+func TestEncodeSameSigWithDifferentTypeShouldBeDifferen(t *testing.T) {
+	sig := unittest.SignatureFixture()
+	encodedAsStaking := EncodeSingleSig(hotstuff.SigTypeStaking, sig)
+	encodedAsRandomBeacon := EncodeSingleSig(hotstuff.SigTypeRandomBeacon, sig)
+	require.NotEqual(t, encodedAsStaking, encodedAsRandomBeacon)
+}
```
Comment (tarakby) [inline consensus/hotstuff/signature/consensus_test.go:49]: The test `TestEncodeDecodeSig` formally includes these 2 tests: it means that these 2 tests are already tested by `TestEncodeDecodeSig`, and in theory are not needed.

**Review by tarakby — APPROVED** [review]
Looks good to me, I left some comments about the tests.

**Review by durkmurder — APPROVED** [review]
LGTM

**Pull request merged**